### PR TITLE
lncli: add command to delegate macaroons

### PR DESCRIPTION
**Pull request by guggero**
This PR addresses #283 and introduces the new command `delegatemacaroon` to the `lncli` tool.

Currently, this just exposes the functionality that the `lncli` already uses internally to the command line interface:
* Take an existing macaroon (by specifying `--macaroonpath`, default is `~/.lnd/admin.macaroon`)
* [optional] Add a timeout to it (by specifying `--timeout`)
* [optional] Restrict it to a specific IP address (by specifying `--ip_address`)

Depending on the `--format` parameter, the following output is produced (truncated for brevity):

**`lncli delegatemacaroon --format=json --timeout=3600 --ip_address=123.123.123.123`**:
```json
{
        "c": [
                {
                        "i": "time-before 2018-04-28T16:02:23.47112259Z"
                },
                {
                        "i": "ipaddr 123.123.123.123"
                }
        ],
        "l": "lnd",
        "i64": "AwoQ7V1quyefYWYwh9TyFCJCBhIBMBoWCgdhZGRyZXNzEgRy....",
        "s64": "BqjN2qrARiy4dRfX7H__3t0XRWqXtVjNs5gyXL7oU9Q"
}
```

**`lncli delegatemacaroon --format=hex --timeout=3600 --ip_address=123.123.123.123`**:
```json
{
        "macaroonHex": "0201036c6e6402bb01030a10ed5d6abb279f61663..."
}
```

Once new first-party caveats are introduced, they can be implemented here as well.

@aakselrod:
Does this cover your idea as described in #283?
As I see it, there are at least two ways to delegate macaroons in `lnd`:
1. Take one of the three existing macaroon files (`admin.macaroon`, `invoice.macaroon` and `readonly.macaroon`) and add first-party caveats to them (as implemented in this PR).
1. Bake fresh macaroons by using the encrypted root key in `macaroons.db` and define custom operations/permissions. Then add first-party caveats to those newly baked macaroons.
The second option has the advantage that the user can customize the allowed operations/permissions. But the big disadvantage that the `macaroons.db` need to be unlocked (`lnd` needs to be stopped) and decrypted with the wallet password.

Am I correct in the assumption that you meant to implement option 1 in your issue, instead of option 2?

Thanks for your feedback!

### CODE REVIEW

**Comment by vegardengen** [comment]
There is definitely a need for option two. For example in the woocommerce LND plugin, there is the need for invoice macaroon rights *plus* some other read-rights. For now, that means we need to give it the admin macaroon, which is not too good security wise.

File: cmd/lncli/commands.go [hunk 1]
```diff
@@ -3160,3 +3162,149 @@ func forwardingHistory(ctx *cli.Context) error {
...
+	Take an existing macaroon (use --macaroonpath) as a template, add
+	restrictions to it and then serialize it either in JSON (--format=json)
+	or binary (--format=binary) format.
+	`,
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name: "format",
+			Usage: "the format to serialize the macaroon in, " +
+				"must be either 'json' or 'binary'",
+			Value: "binary",
```
Comment (aakselrod) [inline cmd/lncli/commands.go]: In testing this, it appears to default to `json` even though the default is set to binary.
```
$ lncli -n simnet delegatemacaroon --ip_address 127.0.0.2     │ing
{                                                                                         │2018-08-28 15:42:34.709 [ERR] DISC: unable to rebroadcast s
        "c": [                                                                            │tale channels: error while retrieving outgoing channels: no
                {                                                                         │ graph edges exist
                        "i": "ipaddr 127.0.0.2"                                           │2018-08-28 15:42:34.816 [INF] CRTR: Filtering chain using 0
                }                                                                         │ channels active
        ],                                                                                │2018-08-28 15:42:34.816 [INF] CRTR: Prune tip for Channel G
        "l": "lnd",                                                                       │raph: height=1500, hash=15b1caeddf87e4046661fe0a50a7445501f
        "i64": "AwoQQp_pVSCfaJXVA-zJtvQq2RIBMBoWCgdhZGRyZXNzEgRyZWFkEgV3cml0ZRoTCgRpbmZvEg│fb8a1e1bbab63afa17ae22525f70b                             
RyZWFkEgV3cml0ZRoXCghpbnZvaWNlcxIEcmVhZBIFd3JpdGUaFgoHbWVzc2FnZRIEcmVhZBIFd3JpdGUaFwoIb2Zm│2018-08-28 15:42:34.821 [INF] CMGR: Server listening on [::
Y2hhaW4SBHJlYWQSBXdyaXRlGhYKB29uY2hhaW4SBHJlYWQSBXdyaXRlGhQKBXBlZXJzEgRyZWFkEgV3cml0ZQ",  │]:9735
        "s64": "ugWq8Ib_KyCLATbzAffsqXMM3UKiJ1rxnIJDialeqhM"                              │2018-08-28 15:42:34.827 [INF] SRVR: Auto peer bootstrapping
}
```

Still working on testing this.

**Comment by aakselrod** [comment]
Tested the new changes (rebased on current master). This LGTM.

**Comment by guggero** [comment]
Now that #1160 is merged, this doesn't offer any new functionality. Closing as redundant.

**Pull request closed without merging**